### DISCUSSION TITLE
Creating Integration Test for QuartzHostedService

### DISCUSSION THREAD

**Discussion by isaacnmartin51** [quartznet/quartznet#1653] [Q&A]
Hi, I am new to using quartz and am curious about how to create integration tests for it.

I am creating the job via dependency injection and embedding it as a hosted service in my web api. 

```
   public static partial class ServiceCollectionExtentions
    {
        public static IServiceCollection ConfigureQuartz(this IServiceCollection services,
            IConfiguration config) =>
            services.Configure<QuartzOptions>(config.GetSection("Quartz"))
                    .AddQuartz(q =>
                    {
                        q.UseMicrosoftDependencyInjectionJobFactory();
                        q.ScheduleJob<ProcessMatchesJob>(trigger => trigger
                            .WithIdentity("ProcessMatchesJobTrigger")
                            .StartNow()
                            .WithSimpleSchedule(x => x
                                .WithIntervalInSeconds(5)
                                .WithRepeatCount(2)));
                    })
                    .AddQuartzHostedService(options =>
                    {
                        options.WaitForJobsToComplete = true;
                    });

    }
```

Startup.cs
` services.ConfigureQuartz(Configuration);`

I thought I could create an integration test to verify that the dependency injection works by creating a `WebApplicationFactory<TEntryPoint>` and checking the services for the hosted service type, but `QuartzHostedService` is internal so I am unable to check for it. 

Test Code
```
        [Test]
        public void ServiceCollection_ShouldContain_Quartz()
        {
            var application = new WebApplicationFactory<Startup>();
            
            using var scope = application.Services.CreateScope();

            var hostedService = scope.ServiceProvider.GetService<QuartzHostedService>();
                       
            Assert.That(hostedService, Is.Not.Null);

        }
```

If I look for the generic `IHostedService` I can see the quartz service, but I feel this is poor test design as I'm reliant on a string rather than a type.

```
        [Test]
        public void ServiceCollection_ShouldContain_Quartz()
        {
            var application = new WebApplicationFactory<Startup>();
            

            using var scope = application.Services.CreateScope();

            var hostedServices = scope.ServiceProvider.GetServices<IHostedService>();

            var quartzService = hostedServices.Where(s => s.ToString().Contains("Quartz")).First();

            Assert.That(quartzService, Is.Not.Null);

        }
        
```
Is this the correct way to test dependency injection with a hosted service when using quartz? Or is there a better practice for integration testing?

**Comment by lahma** [comment]
I'm not sure if you really want to test against the hosted service implementation. Usually in integration test one wants to disable all background services anyway as they cause "jitter". If your main concern is dependency injection might be easier just to create the `ServiceCollection` with your setup and test whether you can resolve every type that implements `IJob`? 

Functional testing for each job is then a different story of course.
